af_ktls.c

```diff
@@ -358,30 +379,569 @@ static void tls_update_senpage_ctx(struct tls_sock *tsk, size_t size)
+				 * the rest of the message. Also, set "early
+				 * eaten" since we've already buffered the skb
+				 * but don't consume yet per tcp_read_sock.
+				 * If function returns 0, does not consume
+				 */
+
+				/* Wait. Why doesn't this code path just set
+				 * eaten? Then tcp_read_sock will eat and
+				 * profit!
+				 */
+				if (!rxm->accum_len) {
+					/* Start RX timer for new message */
+					/*kcm_start_rx_timer(tsk); */
+				}
+
+				tsk->rx_need_bytes = rxm->full_len -
+						       rxm->accum_len;
+				rxm->accum_len += uneaten;
+				rxm->early_eaten = uneaten;
+				desc->count = 0; /* Stop reading socket */
+				break;
+			}
+			rxm->accum_len += uneaten;
+			eaten += uneaten;
+			WARN_ON(eaten != orig_len);
+			break;
+		}
+
+		/* Positive extra indicates ore bytes than needed for the
+		 * message
+		 */
+
+		WARN_ON(extra > uneaten);
+
+
+		ret = decrypt_skb(tsk, head);
+
+		/* Hurray, we have a new message! */
+		tsk->rx_skb_head = NULL;
+		eaten += (uneaten - extra);
+
+		sock_queue_rcv_skb(sk, head);
+	}
+done:
+	if (cloned_orig)
+		kfree_skb(orig_skb);
+	return eaten;
+decryption_fail:
+	kfree_skb(skb);
+	desc->error = -EBADMSG;
+	tsk->rx_skb_head = NULL;
+	desc->count = 0;
+	goto done;
+}
+
+static int tls_tcp_read_sock(struct tls_sock *tsk)
+{
+	read_descriptor_t desc;
+
+	desc.arg.data = tsk;
+	desc.error = 0;
+	desc.count = 1; /* give more than one skb per call */
+
+	/* sk should be locked here, so okay to do tcp_read_sock */
+	tcp_read_sock(tsk->socket->sk, &desc, tls_tcp_recv);
+
+	return desc.error;
+}
+
+static void do_tls_data_ready(struct tls_sock *tsk)
+{
+	int ret;
+
+	ret = tls_tcp_read_sock(tsk);
+	if (ret == -ENOMEM) /* No memory. Do it later */
+		queue_work(tls_wq, &tsk->recv_work);
+	/*queue_delayed_work(tls_wq, &tsk->recv_work, 0); */
+	/* TLS couldn't handle this message. Pass it directly to userspace */
+	else if (ret == -EBADMSG)
+		tls_err_abort(tsk);
+
+}
+
+/* Called with lower socket held */
 static void tls_data_ready(struct sock *sk)
 {
 	struct tls_sock *tsk;
 
-	xprintk("--> %s", __FUNCTION__);
+	xprintk("--> %s", __func__);
 
 	read_lock_bh(&sk->sk_callback_lock);
-
+	//TODO: forgot to lock tsk?
 	tsk = (struct tls_sock *)sk->sk_user_data;
-	if (unlikely(!tsk || tsk->rx_stopped)) {
+	if (unlikely(!tsk || tsk->rx_stopped))
 		goto out;
+
+	if (tsk->rx_need_bytes) {
+		if (tcp_inq(sk) >= tsk->rx_need_bytes)
+			tsk->rx_need_bytes = 0;
+		else
+			goto out;
 	}
-	queue_work(tls_wq, &tsk->recv_work);
 
-  out:
+	do_tls_data_ready(tsk);
+
+out:
 	read_unlock_bh(&sk->sk_callback_lock);
 }
 
+#include "dtls-window.c"
+
+//Loop through the SKBs. Decrypt each one and, if valid, add it to recv queue
+static int dtls_udp_read_sock(struct tls_sock *tsk)
+{
+	struct sk_buff *p, *next, *skb;
+	int ret = 0;
+
+	skb_queue_walk_safe(&tsk->socket->sk->sk_receive_queue, p, next) {
+
+		ssize_t len;
+		struct tls_rx_msg *rxm;
+
+		rxm = tls_rx_msg(p);
+		memset(rxm, 0, sizeof(*rxm));
```

Comment:
We are placing rxm in the skb->cb, which are used in the lower layers. Memory there may contain non-zero values